raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -584,211 +721,244 @@ private void updateVoterConnections(FindQuorumResponseData response) {
 
-    private boolean hasInconsistentLeader(int epoch, OptionalInt leaderId) {
+    private boolean hasConsistentLeader(int epoch, OptionalInt leaderId) {
         // Only elected leaders are sent in the request/response header, so if we have an elected
         // leaderId, it should be consistent with what is in the message.
-        if (epoch != quorum.epoch()) {
-            return false;
-        } else {
-            if (!quorum.leaderId().isPresent())
-                return false;
-            if (!leaderId.isPresent())
-                return false;
-            return !quorum.leaderId().equals(leaderId);
-        }
-    }
-
-    private void becomeFollower(int epoch, OptionalInt leaderId) throws IOException {
-        if (leaderId.isPresent()) {
-            becomeFollower(leaderId.getAsInt(), epoch);
+        if (leaderId.isPresent() && leaderId.getAsInt() == quorum.localId) {
+            // The response indicates that we should be the leader, so we verify that is the case
+            return quorum.isLeader();
         } else {
-            becomeUnattachedFollower(epoch);
+            return epoch != quorum.epoch()
+                || !leaderId.isPresent()
+                || !quorum.leaderId().isPresent()
+                || leaderId.equals(quorum.leaderId());
         }
     }
 
-    private boolean maybeHandleError(RaftResponse.Inbound response, long currentTimeMs) throws IOException {
-        ConnectionCache.ConnectionState connection = connections.getOrCreate(response.sourceId());
-
-        final ApiMessage data = response.data();
-        final Errors responseError;
-        final int responseEpoch;
-        final OptionalInt responseLeaderId;
-
-        if (data instanceof FetchQuorumRecordsResponseData) {
-            FetchQuorumRecordsResponseData fetchRecordsResponse = (FetchQuorumRecordsResponseData) data;
-            responseError = Errors.forCode(fetchRecordsResponse.errorCode());
-
-            if (responseError == Errors.OFFSET_OUT_OF_RANGE) {
-                connection.onResponseReceived(response.correlationId);
-                return false;
+    /**
+     * Handle response errors that are common across request types.
+     *
+     * @param error Error from the received response
+     * @param leaderId Optional leaderId from the response
+     * @param epoch Epoch received from the response
+     * @return Optional value indicating whether the error was handled here and the outcome of
+     *    that handling. Specifically:
+     *
+     *    - Optional.empty means that the response was not handled here and the custom
+     *        API handler should be applied
+     *    - Optional.of(true) indicates that the response was successfully handled here and
+     *        the request does not need to be retried
+     *    - Optional.of(false) indicates that the response was handled here, but that the request
+     *        will need to be retried
+     */
+    private Optional<Boolean> maybeHandleStandardError(
```

Comment:
nit: my paranoid hat is urging me to suggest rename it to `maybeHandleCommonResponse` to be aligned with `validateLeaderOnlyRequest / validateVoterOnlyRequest`.

Also, at the end could we check that if the `error` passed in is NONE then we can just return `Optional.of(true)` and hence the caller do not need to check this anymore?

Reply:
More generally speaking, I think the contract should be `if there's any issues indicated in the other fields, then the error field should always not be NONE, though vice versa may not be true`, and the error handling logic can be simpler.

Reply:
Ack on the naming. I am not sure I follow your other point though. Having no error is a necessary, but insufficient condition to trigger the rest of the response handling logic. The intended semantic is `Optional.of(value)` means we handled it here and we don't need to bother with any additional handling. On the other hand, `Optional.empty()` means that all basic validations passed and we need to handle the case with the respective API semantics in mind.

I realize this may fall short of perfection, but I think it is a reasonable compromise to ensure consistent handling of common cases.